server/src/graql/gremlin/RelationTypeInference.java

```diff
@@ -0,0 +1,184 @@
+        Multimap<Variable, Type> instanceVarTypeMap = HashMultimap.create();
+        int oldSize;
+        do {
+            oldSize = instanceVarTypeMap.size();
+            allFragments.stream()
+                    .filter(fragment -> labelVarTypeMap.containsKey(fragment.start())) // restrict to types
+                    .filter(fragment -> fragment instanceof InIsaFragment || fragment instanceof InSubFragment) //
+                    .forEach(fragment -> instanceVarTypeMap.put(fragment.end(), labelVarTypeMap.get(fragment.start())));
+        } while (oldSize != instanceVarTypeMap.size());
+        return instanceVarTypeMap;
+    }
+
+    // find all vars representing types
+    private static Map<Variable, Type> getLabelVarTypeMap(TransactionOLTP tx, Set<Fragment> allFragments) {
+        Map<Variable, Type> labelVarTypeMap = new HashMap<>();
+        allFragments.stream()
+                .filter(LabelFragment.class::isInstance)
+                .forEach(fragment -> {
+                    // TODO: labels() should return ONE label instead of a set
+                    SchemaConcept schemaConcept = tx.getSchemaConcept(
+                            Iterators.getOnlyElement(((LabelFragment) fragment).labels().iterator()));
+                    if (schemaConcept != null && !schemaConcept.isRole() && !schemaConcept.isRule()) {
+                        labelVarTypeMap.put(fragment.start(), schemaConcept.asType());
+                    }
+                });
+        return labelVarTypeMap;
+    }
+
+    private static Multimap<Variable, Variable> getRelationRolePlayerMap(
+            Set<Fragment> allFragments, Multimap<Variable, Type> instanceVarTypeMap) {
+        // relation vars and its role player vars
+        Multimap<Variable, Variable> relationRolePlayerMap = HashMultimap.create();
+        allFragments.stream().filter(OutRolePlayerFragment.class::isInstance)
+                .forEach(fragment -> relationRolePlayerMap.put(fragment.start(), fragment.end()));
+
+        // find all the relation requiring type inference
+        Iterator<Variable> iterator = relationRolePlayerMap.keySet().iterator();
+        while (iterator.hasNext()) {
+            Variable relation = iterator.next();
+
+            // the relation should have at least 2 known role players so we can infer something useful
+            if (instanceVarTypeMap.containsKey(relation) ||
+                    relationRolePlayerMap.get(relation).size() < 2) {
+                iterator.remove();
+            } else {
+                int numRolePlayersHaveType = 0;
+                for (Variable rolePlayer : relationRolePlayerMap.get(relation)) {
+                    if (instanceVarTypeMap.containsKey(rolePlayer)) {
+                        numRolePlayersHaveType++;
+                    }
+                }
+                if (numRolePlayersHaveType < 2) {
+                    iterator.remove();
+                }
+            }
+        }
+        return relationRolePlayerMap;
```

Comment:
when possible it's better to avoid this pattern of creating a collection and then retrieve the iterator and manually remove what's not needed, we have many tools that can help us building the correct map in one go

Reply:
Implemented in QP refactor 2